sotoki/sotoki.py

```diff
@@ -890,13 +882,153 @@ def get_response_headers(url):
     return "default", "default"
 
 
-def download_image(url, fullpath, convert_png=False, resize=False):
+def get_redirection_to_duplicate_or_skip(url, convert_png, resize, ext):
+    """ check the URL for various types of duplicates and return a tuple containing
+        name of the file to which image can be redirected and whether we can skip the download """
+
+    def download_and_return_name(url, source, convert_png, resize, ext):
+        """ genetare the file name given the parameters, check if present in fs and download if not present
+            returns the name of the file which was generated from the params """
+
+        convertion = "png" if convert_png else "org"
+        size = "org" if not resize else str(resize)
+        image_path = (
+            pathlib.Path(output_dir)
+            .joinpath("common_images")
+            .joinpath(f"{source}_{convertion}_{size}{ext}")
+        )
+
+        # download the duplicate file only once
+        if not image_path.exists():
+            download_image(
+                url=url,
+                fullpath=str(image_path),
+                convert_png=convert_png,
+                resize=resize,
+                skip_duplicate_check=True,
+            )
+        return image_path.name
+
+    parsed_url = urllib.parse.urlparse(url)
+    if "gravatar.com" in parsed_url.netloc:
+        # parse the url
+        url_parts = list(parsed_url)
+        query = dict(urllib.parse.parse_qsl(url_parts[4]))
+
+        # check if autogenerated identicon from gravatar
+        # for more details see https://en.gravatar.com/site/implement/images/
+        if query.get("d", query.get("default", "")) == "identicon" and query.get(
+            "f", query.get("forcedefault", "")
+        ) in ["y", "1"]:
+            return None, True
+
+        # check for mystery-person type identicon
+        if query.get("d", query.get("default", "")) == "mp":
+            return (
+                download_and_return_name(
+                    url, "gravatar_mystery", convert_png, resize, ext
+                ),
+                False,
+            )
+
+        # check for blank identicon
+        if query.get("d", query.get("default", "")) == "blank":
+            return (
+                download_and_return_name(
+                    url, "gravatar_blank", convert_png, resize, ext
+                ),
+                False,
+            )
+
+        # update query with size = 128 (its the size in which we download identicons)
+        query.update({"s": "128"})
+        url_parts[4] = urllib.parse.urlencode(query)
+        url_with_size = urllib.parse.urlunparse(url_parts)
+
+        # check if potentially a duplicate default image at size = 128
+        headers = get_response_headers(url_with_size)
+        if headers.get("content-length") == "4268":
```

Comment:
Where's this number coming from? How do we keep it updated overtime?
We should add a comment here

Reply:
These content-length values are the content-length value for the two types of default images, the default1 being the blue 'G' logo and default2 one being the green autogenerated identicon if hash is invalid these need to be manually changed if the images change but they are here for a pretty long time and shall stay to be so. I've put this comment there.

Reply:
Can you add the URL of such default icon?

Reply:
Okay, I didn't read the comment, sorry. Here's a link to both of them -
https://www.gravatar.com/avatar/wrong_hash?s=128
https://www.gravatar.com/avatar/XXXXXXXXXXXXXXXXXXXXXXXXX?s=128&d=identicon

Will add them in the code whenever I push a new commmit.

Reply:
pushed it to master